### PR TITLE
Added TimerX::isChannelConfiguredAsInput.

### PR DESCRIPTION
**Pull request by dhebbeker**
This function is especially useful in the ISR in order to check the interrupt source.

#### To do:
- [x] squash all commits into one
- [x] do a spot-check of a few different families RMs

### CODE REVIEW

File: src/modm/platform/timer/stm32/general_purpose.cpp.in [hunk 1]
```diff
@@ -172,3 +172,31 @@ modm::platform::Timer{{ id }}::enableInterruptVector(bool enable, uint32_t prior
...
+	switch (channel) {
+		case 1:
+			isInput = TIM{{ id }}->CCMR1 bitand TIM_CCMR1_CC1S;
+			break;
+		case 2:
+			isInput = TIM{{ id }}->CCMR1 bitand TIM_CCMR1_CC2S;
+			break;
+%% if (id <= 5)
+		case 3:
+			isInput = TIM{{ id }}->CCMR2 bitand TIM_CCMR2_CC3S;
```
Comment (salkinium) [inline src/modm/platform/timer/stm32/general_purpose.cpp.in]: Please use `&` instead of `bitand`. The idea is to visually distinguish bit operations and logic operations by using `&`, `|`, etc for bit operations and `and`, `or`, etc for logic operations. `bitand` kinda messes this up.

**Comment by salkinium** [comment]
How would this interact with the interrupt flags? Is there a way to combine the two smartly? ("clear flag only if input" or something useful like that)

**Comment by rleh** [comment]
Aren't there Timers with more than 4 channels?

**Comment by dhebbeker** [comment]
> Aren't there Timers with more than 4 channels?

Can you recommend a good resource to look it up? Is there a more efficient way than reading all applicable references manuals?

**Comment by dhebbeker** [comment]
> 
> 
> How would this interact with the interrupt flags? Is there a way to combine the two smartly? ("clear flag only if input" or something useful like that)

@salkinium: I use this function in a generic IRQ handler in order to differentiate the interrupt source. In case a capture/compare interrupt flag is set and the corresponding interrupt is enabled *then* I check if the corresponding channel is configured as input (`isChannelConfiguredAsInput()`) to call the corresponding handler to input events *or* call the handler for output compare events. `acknowledgeInterruptFlags()` is called independently of this last differentiation. This is similar to the [procedure](https://github.com/MaJerle/stm32f429/blob/5ca39ed098d994fd645f35aec15532aafeff9fac/00-STM32F4xx_HAL_DRIVERS/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_tim.c#L2787) in the STM HAL drivers (which shall not justify it).

In general this function could be used to make decisions outside of an interrupt.

**Comment by salkinium** [comment]
> Can you recommend a good resource to look it up? Is there a more efficient way than reading all applicable references manuals?

Nope, but Timer instances are basically the same across all STM32. So best do a spot-check of a few different families RMs. I cannot recall seeing more than four channels per Timer, but I haven't looked too closely either.

**Comment by salkinium** [comment]
(please squash your commits into one)

**Comment by dhebbeker** [comment]
In order to check if there are STM32 MCUs with more than 4 channels per timer I did a search for the following regular expressions in the [CMSIS](https://github.com/modm-io/cmsis-header-stm32/) files:

- `TIM_\S+_CC1S\b`: 117 matches, all `TIM_CCMR1_CC1S`
- `TIM_\S+_CC2S\b`: 117 matches, all `TIM_CCMR1_CC2S`
- `TIM_\S+_CC3S\b`: 117 matches, all `TIM_CCMR2_CC3S`
- `TIM_\S+_CC4S\b`: 117 matches, all `TIM_CCMR2_CC4S`
- ***`TIM_\S+_CC5S\b`: 0 matches***
- **`TIM_CCMR3_CC`: 0 matches**
- `TIM_CCR1\b`: 117 machtes
- `TIM_CCR2\b`: 117 machtes
- `TIM_CCR3\b`: 117 machtes
- `TIM_CCR4\b`: 117 machtes
- **`TIM_CCR5\b`: 76 machtes**
- **`TIM_CCR6\b`: 76 machtes**
- **`TIM_CCR7\b`: 0 machtes**
- `TIM_CCMR1\b`: 117 matches
- `TIM_CCMR2\b`: 117 matches
- **`TIM_CCMR3\b`: 76 matches**
- **`TIM_CCMR1\b`: 0matches**

I also checked RM for [STM32H742, STM32H743/753 and STM32H750 Value line](https://www.st.com/resource/en/reference_manual/dm00314099.pdf) and [STM32L4x5 and STM32L4x6 advanced Arm®-based 32-bit MCUs](https://www.st.com/resource/en/reference_manual/dm00083560.pdf): TIM1/TIM8 Up to **6** independent channels, but input capture not with channels 5, 6. → implementation still valid for those MCUs

In summary: There are some STM32 MCU which have up to 6 timer channels. But only up to 4 input capture channels. Thus I am confident, that the implementation does not lack the consideration of settings of capture input channels.

**Review by salkinium — APPROVED** [review]
Thanks!

**Comment by rleh** [comment]
> > Aren't there Timers with more than 4 channels?
>
> Can you recommend a good resource to look it up? Is there a more efficient way than reading all applicable references manuals?

Regular expression search in ST/CMSIS headers, as you have already done, is a quick and easy way.

Half a year ago I started to do some analysis on ST reference manuals: https://github.com/rleh/st-data-crawler
The scripts crawls all reference manuals, split them into chapters and group similar chapters, e.g.:

<details>
<summary>
<pre>
Similar ADC peripherals: 
 [ 4] RM0432,  95S: STM32L4Rxxx and STM32L4Sxxx advanced Arm®-based 32-bit MCUs
 [ 9] RM0440, 123S: STM32G4 Series advanced Arm®-based 32-bit MCUs
 [...]
[...]
</pre>
</summary>
<pre>
Similar ADC peripherals: 
        [ 4] RM0432,  95S: STM32L4Rxxx and STM32L4Sxxx advanced Arm®-based 32-bit MCUs
        [ 9] RM0440, 123S: STM32G4 Series advanced Arm®-based 32-bit MCUs
        [12] RM0394, 113S: STM32L41xxx/42xxx/43xxx/44xxx/45xxx/46xxx advanced Arm®-based 32-bit MCUs
        [15] RM0438, 113S: STM32L552xx and STM32L562xx advanced Arm®-based 32-bit MCUs
        [16] RM0434,  89S: Multiprotocol wireless 32-bit MCU Arm®-based Cortex®-M4 with FPU, Bluetooth® Low-Energy and 802.15.4 radio solution
        [18] RM0392, 115S: STM32L4x1 advanced ARM®-based 32-bit MCUs
        [22] RM0316, 109S: STM32F303xB/C/D/E, STM32F303x6/8, STM32F328x8, STM32F358xC, STM32F398xE advanced ARM®-based MCUs
        [23] RM0366,  85S: STM32F301x6/8 and STM32F318x8 advanced ARM®-based 32-bit MCUs
        [34] RM0351, 116S: STM32L4x5 and STM32L4x6 advanced Arm®-based 32-bit MCUs
        [35] RM0365, 106S: STM32F302xB/C/D/E and STM32F302x6/8 advanced ARM®-based 32-bit MCUs
Similar ADC peripherals: 
        [ 1] RM0033,  44S: STM32F205xx, STM32F207xx, STM32F215xx and STM32F217xx advanced ARM-based 32-bit MCUs
        [ 2] RM0041,  28S: STM32F100xx advanced ARM®-based 32-bit MCUs
        [10] RM0410,   0S: STM32F76xxx and STM32F77xxx advanced Arm®-based 32-bit MCUs
        [13] RM0399,   0S: STM32H745/755 and STM32H747/757 advanced Arm®-based 32-bit MCUs
        [14] RM0433,   0S: STM32H742, STM32H743/753 and STM32H750 Value line advanced Arm®-based 32-bit MCUs
        [19] RM0008,  39S: STM32F101xx, STM32F102xx, STM32F103xx, STM32F105xx and STM32F107xx advanced Arm®-based 32-bit MCUs
        [20] RM0368,  30S: STM32F401xB/C and STM32F401xD/E advanced Arm®-based 32-bit MCUs
        [24] RM0383,  30S: STM32F411xC/E advanced Arm®-based 32-bit MCUs
        [26] RM0364,   0S: STM32F334xx advanced Arm®-based 32-bit MCUs
        [28] RM0090,  45S: STM32F405/415, STM32F407/417, STM32F427/437 and STM32F429/439 advanced Arm®-based 32-bit MCUs
        [29] RM0360,   0S: STM32F030x4/x6/x8/xC and STM32F070x6/xB advanced ARM®-based 32-bit MCUs
        [32] RM0038,  46S: STM32L100xx, STM32L151xx, STM32L152xx and STM32L162xx advanced ARM®-based 32-bit MCUs
Similar ADC peripherals: 
        [ 0] RM0367,  50S: Ultra-low-power STM32L0x3 advanced Arm®-based 32-bit MCUs
        [ 3] RM0444,  61S: STM32G0x1 advanced Arm®-based 32-bit MCUs
        [ 7] RM0454,  61S: STM32G0x0 advanced Arm®-based 32-bit MCUs
        [11] RM0451,  47S: Ultra-low-power STM32L0x0 advanced Arm®-based 32-bit MCUs
        [21] RM0376,  49S: Ultra-low-power STM32L0x2 advanced Arm®-based 32-bit MCUs
        [27] RM0377,  50S: Ultra-low-power STM32L0x1 advanced Arm®-based 32-bit MCUs
        [36] RM0091,  41S: STM32F0x1/STM32F0x2/STM32F0x8 advanced ARM®-based 32-bit MCUs
Similar ADC peripherals: 
        [ 5] RM0430,  30S: STM32F413/423 advanced Arm®-based 32-bit MCUs
        [ 6] RM0431,  47S: STM32F72xxx and STM32F73xxx advanced Arm®-based 32-bit MCUs
        [ 8] RM0402,  31S: STM32F412 advanced Arm®-based 32-bit MCUs
        [17] RM0385,  48S: STM32F75xxx and STM32F74xxx advanced Arm®-based 32-bit MCUs
        [25] RM0401,  28S: STM32F410 advanced Arm®-based 32-bit MCUs
        [30] RM0390,  47S: STM32F446xx advanced Arm®-based 32-bit MCUs
        [31] RM0386,  46S: STM32F469xx and STM32F479xx advanced Arm®-based 32-bit MCUs
        [33] RM0313,  27S: STM32F37xxx advanced Arm®-based 32-bit MCUs
</pre>
</details>

Does not yet work with all peripherals and there are many features not yet implemented, such as displaying the chapters for a specific peripheral from all reference manuals in a Side-by-side-PDF-diff-viewer.
